flocker/node/agents/ebs.py

```diff
@@ -997,23 +1019,16 @@ def _attach_ebs_volume(self, volume_id, instance_id, device):
         return volume.attach_to_instance(
             InstanceId=instance_id, Device=device)
 
-    def _next_device(self, instance_id, volumes, devices_in_use):
+    def _next_device(self, devices_in_use):
         """
-        Get the next available EBS device name for a given EC2 instance.
-
-        Algorithm:
-        1. Get all ``Block devices`` currently in use by given instance:
-            a) List all volumes visible to this instance.
-            b) Gather device IDs of all devices attached to (a).
-        2. Devices available for EBS volume usage are ``/dev/sd[f-p]``.
-           Find the first device from this set that is currently not
-           in use.
+        Get the next available EBS device name for this EC2 instance.
+
+        Devices available for EBS volume usage are ``/dev/sd[f-p]``.
+        Find the first device from this set that is currently not
+        in use.
         XXX: Handle lack of free devices in ``/dev/sd[f-p]`` range
         (see https://clusterhq.atlassian.net/browse/FLOC-1887).
 
-        :param unicode instance_id: EC2 instance ID.
-        :param volumes: Collection of currently known
-            ``BlockDeviceVolume`` instances.
         :param set devices_in_use: Unicode names of devices that are
```

Comment:
Hmm, yes that's correct, these come in from the previous returned results of `_next_device` if an attach failed. So yes, I agree with your logic, we can drop this parameter entirely and the use of `ignored_devices` in the `attach_volume` method. Intriguingly, it looks to me like `ignored_devices` might not have worked properly even prior to this branch.